text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+  readonly enableECSManagedTags?: boolean;
+  readonly deploymentController?: DeploymentController;
+}
+
+export interface BaseServiceProps extends BaseServiceOptions {
+  readonly launchType: LaunchType;
+}
+```
+
+As before, the Go translation would have the getter methods (e.g. readonly cluster) defined in a the generated interface, which correspond to
+non-exported struct members in the struct that implements each getter.
+
+```go
+package ecs
+
+type BaseServiceOptionsIface interface {
+    Cluster()                ICluster
+    DesiredCount()           *int
+    ServiceName()            *string
+    MaxHealthyPercent()      *int
+    MinHealthyPercent()      *int
+    HealthCheckGracePeriod() *Duration
+    CloudMapOptions()        CloudMapOptions
+    PropagateTags()          PropagatedTagSource
+    EnableECSManagedTags()   *bool
+    DeploymentController()   DeploymentController
+}
+
+type BaseServicePropsIface interface {
+    BaseServiceOptionsIface  // embeddeded interface
+    LaunchType()             LaunchType
+}
+
+type BaseServiceOptions struct {
+    cluster                ICluster
+    desiredCount           *int
+    serviceName            *string
+    maxHealthyPercent      *int
+    minHealthyPercent      *int
+    healthCheckGracePeriod *Duration
+    cloudMapOptions        CloudMapOptions
+    propagateTags          PropagatedTagSource
+    enableECSManagedTags   *bool
+    deploymentController   DeploymentController
+}
+
+type BaseServiceProps struct {
+    BaseServiceOptions  // embedded (anonymous) field
+    launchType          LaunchType
+}
+
+func (o BaseServiceOptions) Cluster() ICluster { return o.cluster }
+func (o BaseServiceOptions) DesiredCount() int { return o.desiredCount }
+
+// ... etc
+
+```
+
+This allows the embedding interface (here, `BaseServicePropsIFace`) to "inherit" all the methods defined in the embedded interface
+(`BaseServiceOptionsIFace`). Similarly, the embedding struct (`BaseServiceProps`) would have access to all the properties of the embedded struct
+(`BaseServiceOptions`) through having it as an anonymous field (See: [Go playground example](https://play.golang.org/p/XRpTH3LN6bx)).
+
+```go
+props := BaseServiceProps{
+    BaseServiceOptions{
+      serviceName: "myService",
+      ...
+    },
+    launchType: "EC2",
+}
+
+props.serviceName  // => "myService"
+```
+
+### Notes/Concerns
+
+* When fields in the Go struct are exported, the struct cannot have a method with the same name (e.g. exported field `Bar` , and method `Bar()`.
+   Using a `Bar()` getter on a private field bar is most consistent with idiomatic Go;  however, this may make code generation more difficult since it
+would potentially necessitate converting public properties on the source (JSII) class into an interface implemented by the custom struct (Go “class”).
+The *alternative* to this is to prefix the methods in the generated Go interface with Get, e.g. `GetBar()` to avoid the name collision.
```

Comment:
I don't understand what this section is trying to say.

In a struct, all properties are always public, and they are always readonly, so the shape of the interface has to (more or less) be this:

```go
type StructIface interface {
  Foo() int;
  Bar() string;
}
```

Is the question how to make the fields of the struct mutable? The JSII contract doesn't say anything about how structs ought to be filled by users. It can be through declaring and mutating the struct fields directly, or it can be through a second interface, at the binding author's discretion (though fully formed structs are in most languages treated as immutable, so mutability is possible but deviating from the norm).

Or is this concern not about structs but about regular/behavioral/IResource interfaces? Where we could have a property that is mutable, and so needs both a getter and a setter, and what do we call those?

Reply:
Not exactly -- I realized that this point is in the wrong place because it refers to the translation of classes, not interfaces.

If I have  a typescript class defined with a readonly property, e.g.:

```ts
class Struct {
  readonly foo: string;
}
```

Then the literal way of translating it into a go struct would be with an exported member `Foo`, e.g.:
```go
type Struct struct {
	Foo string
}
```

However, the idiomatic way for accessing properties mentioned here is to define a Getter method (e.g. `Foo()` ). 

If I define the following interface with the `Foo()` getter method:
```go
type StructIface interface {
	Foo() string
}
```

Go will not allow a struct that has an exported property with the same name as an interface's method to implement that interface:

```
func (s Struct) Foo() string {
	return s.Foo
} 
s := Struct{"foo"}
fmt.Printf(s.Foo) // => type Struct has both field and method named Foo
```

See: https://play.golang.org/p/LEbI4-GCMfG


To get around that, we would have to translate the readonly properties of a typescript class into private members of a struct in go:

```go
type Struct struct {
	foo string
}

func (s Struct) Foo() string {
	return s.foo
} 
```

https://play.golang.org/p/SbWwrjq_Mqk

Doing it this way would mean for every class, we would have to generate an interface that just holds the getter methods for public properties of a class in addition to the struct.

The alternative mentioned would be to keep the property as exported and generated an interface with the method `GetFoo()` instead, but that seems a little clunky to me (see: https://play.golang.org/p/9UTTLDX3VV7).

Reply:
okay I lied we're probably doing `GetFoo()`.

Reply:
Oh god.

Might I suggest:

```go
type SomeClass struct {
	foo string
}

func (s SomeClass) Foo() string {
	return s.foo
}

func (s SomeClass) SetFoo(string value)  {
	s.foo = value
} 
```

Instead?

Reply:
This would make instantiating `SomeClass` from a different package look like this:

```go
var x SomeClass
x.SetFoo("foovalue")
```
...for every single property, rather than simply:

```go
x := SomeClass{
    foo: "foovalue"
}
```
(I'm not happy about the `GetFoo` thing either =P)